jerry-core/parser/js/js-lexer.c

```diff
@@ -1280,15 +1280,37 @@ lexer_parse_string (parser_context_t *context_p, /**< context */
  */
 static inline void
 lexer_parse_octal_number (parser_context_t *context_p, /** context */
-                          const uint8_t **source_p) /**< current source position */
+                          const uint8_t **source_p, /**< current source position */
+                          bool is_legacy) /**< is legacy octal number */
 {
-  do
+#if !ENABLED (JERRY_ESNEXT)
+  JERRY_UNUSED (is_legacy);
+#endif /* !ENABLED (JERRY_ESNEXT) */
+  while (true)
   {
-    (*source_p)++;
+    do
+    {
+      (*source_p)++;
+    }
+    while (*source_p < context_p->source_end_p
+           && *source_p[0] >= LIT_CHAR_0
+           && *source_p[0] <= LIT_CHAR_7);
+#if ENABLED (JERRY_ESNEXT)
+    if (*source_p < context_p->source_end_p && *source_p[0] == LIT_CHAR_UNDERSCORE)
+    {
+      (*source_p)++;
+      if (is_legacy ||
+          *source_p  == context_p->source_end_p ||
+          *source_p[0] < LIT_CHAR_0 ||
+          *source_p[0] > LIT_CHAR_7)
+      {
+        parser_raise_error (context_p, PARSER_ERR_INVALID_UNDERSCORE_USE_IN_NUMBER);
+      }
+      continue;
+    }
+#endif /* ENABLED (JERRY_ESNEXT) */
```

Comment:
This function is almost identical to `lexer_check_underscore`. Cannot we combine it?